crates/miden-lib/asm/miden/contracts/auth/multisig.masm

```diff
@@ -1,24 +1,273 @@
+
+            padw mem_loadw.TEMP_PK
+            # => [PK, MSG, MSG, i]
+
+            swapw
+            # => [MSG, PK, MSG, i]
+
+            # Emit the authentication request event that pushes a signature for the message to the advice stack.
+            emit.AUTH_REQUEST
+
+            swapw
+            # => [PK, MSG, MSG, i ]
+
+            exec.rpo_falcon512::verify
+            # => [MSG, i]
+
+            mem_load.SUCCESSFUL_VERIFICATIONS
+            add.1
+            mem_store.SUCCESSFUL_VERIFICATIONS
+            # => [MSG, i]
+
+        else
+            dropw
+            # => [MSG, i]
+        end
+        # => [MSG, i]
+
+        movup.4
+        # => [i, MSG]
+
+        # Decrement counter and check if we should continue looping
+        sub.1
+        # => [i-1, MSG]
+
+        dup neq.0
+        # => [should_continue, i-1, MSG]
+    end
+    # => [i-1, MSG]
+
+    # Check number of successful signature verifications
+    # -----------------------------------------------------------------------------------------
+
+    drop
+    # => [MSG]
+
+    mem_load.SUCCESSFUL_VERIFICATIONS dup
+    # => [num_verifications, num_verifications, MSG]
+
+    eq.0 swap
+    # => [num_verifications, is_zero_verifications, MSG]
+
+    mem_load.THRESHOLD
+    # => [threshold, num_verifications, is_zero_verifications, MSG]
+
+    gte not or
+    # => [is_unauthorized, MSG]
+
+    # If signatures are non-existent the tx will fail here.
+    if.true
+        padw mem_loadw.TEMP_PK swapw
+        # => [MSG, SIG_KEY]
+
+        emit.UNAUTHORIZED_EVENT
+    end
+    # => [MSG, SIG_KEY]
+
+    dropw dropw
+    # => []
+end
+
+#! Multi Signature Authentication Component
+#!
+#! OS:  [SALT]
+#! AS Map => {h(SIG_0, MSG): SIG_0, h(SIG_1, MSG): SIG_1, h(SIG_n, MSG): SIG_N}
+#!
+#! Outputs: []
+# main entrypoint ->
 export.auth__tx_rpo_falcon512_multisig
-    # => [TX_HASH, pad(16)]
-    dropw
+    exec.account::incr_nonce drop
+    # => [SALT]
+
+    # Computing transaction summary
+    # ______________________________________
+
+    exec.auth::create_tx_summary
+    # => [SALT, OUTPUT_NOTES_COMMITMENT, INPUT_NOTES_COMMITMENT, ACCOUNT_DELTA_COMMITMENT]
+
+    exec.auth::adv_insert_hqword
+    # => [SALT, OUTPUT_NOTES_COMMITMENT, INPUT_NOTES_COMMITMENT, ACCOUNT_DELTA_COMMITMENT]
+
+    # the commitment to the tx summary is the message that is signed
+    exec.auth::hash_tx_summary
+    # => [TX_SUMMARY_COMMITMENT]
+
+    mem_storew.MESSAGE dropw
+    # => []
+
+    # Verifying owner signatures
+    # ______________________________________
+
+    push.THRESHOLD_SLOT
+    # => [index]
+
+    exec.account::get_item
+    # => [0, 0, 0, num_required_signatures]
+
+    drop drop drop
+    # => [num_required_signatures]
+
+    exec.verify_signatures_loop
+    # => []
+
+    # Writing executed transaction MSG to map
+    # ______________________________________
+
+    mem_loadw.MESSAGE
+    # => [MSG]
+
+    exec.replay_protection
+    # => []
```

Comment:
Can we break replay_protection in two part, one is to check whether the MSG exist in the map which can be called before the verification of signature and another is to add the MSG in the map which will called in the last.
It is  the extra optimisation.